controllers/complianceeventsapi/auth.go

```diff
@@ -0,0 +1,194 @@
+	return strings.TrimSpace(strings.TrimPrefix(req.Header.Get("Authorization"), "Bearer"))
+}
+
+// getClientFromToken will generate a Kubernetes client using the input config and token. No authentication data from
+// the input config is used in the generated client.
+func getClientFromToken(cfg *rest.Config, token string) (*kubernetes.Clientset, error) {
+	userConfig := &rest.Config{
+		Host:    cfg.Host,
+		APIPath: cfg.APIPath,
+		TLSClientConfig: rest.TLSClientConfig{
+			CAFile:     cfg.TLSClientConfig.CAFile,
+			CAData:     cfg.TLSClientConfig.CAData,
+			ServerName: cfg.TLSClientConfig.ServerName,
+			Insecure:   cfg.TLSClientConfig.Insecure,
+		},
+		BearerToken: token,
+	}
+
+	return kubernetes.NewForConfig(userConfig)
+}
+
+// canRecordComplianceEvent will perform certificate or token authentication and perform a subject access review to
+// ensure the input user has patch access to patch the policy status in the managed cluster namespace. An error is
+// returned if the authorization could not be determined. Note that authenticatedClient and authenticator can be nil
+// if certificate authentication isn't used. If both certificate and token authentication is present, certificate takes
+// precedence.
+func canRecordComplianceEvent(
+	cfg *rest.Config,
+	authenticatedClient *kubernetes.Clientset,
+	authenticator *apiserverx509.Authenticator,
+	clusterName string,
+	req *http.Request,
+) (bool, error) {
+	postRules := authzv1.ResourceAttributes{
+		Group:       "policy.open-cluster-management.io",
+		Version:     "v1",
+		Resource:    "policies",
+		Verb:        "patch",
+		Namespace:   clusterName,
+		Subresource: "status",
+	}
+
+	// req.TLS.PeerCertificates will be empty if certificate authentication is not enabled (e.g. endpoint is not HTTPS)
+	if req.TLS != nil && len(req.TLS.PeerCertificates) > 0 {
+		resp, ok, err := authenticator.AuthenticateRequest(req)
+		if err != nil {
+			if errors.As(err, &x509.UnknownAuthorityError{}) || errors.As(err, &x509.CertificateInvalidError{}) {
+				return false, ErrNotAuthorized
+			}
+
+			return false, err
+		}
+
+		if !ok {
+			return false, ErrNotAuthorized
+		}
+
+		review, err := authenticatedClient.AuthorizationV1().SubjectAccessReviews().Create(
+			req.Context(),
+			&authzv1.SubjectAccessReview{
+				Spec: authzv1.SubjectAccessReviewSpec{
+					ResourceAttributes: &postRules,
+					User:               resp.User.GetName(),
+					Groups:             resp.User.GetGroups(),
+					UID:                resp.User.GetUID(),
+				},
+			},
+			metav1.CreateOptions{},
+		)
+		if err != nil {
+			return false, err
+		}
+
+		if !review.Status.Allowed {
+			log.V(0).Info(
+				"The user is not authorized to record a compliance event",
+				"cluster", clusterName,
+				"user", resp.User.GetName(),
+			)
+		}
+
+		return review.Status.Allowed, nil
+	}
+
+	token := parseToken(req)
+	if token == "" {
+		return false, ErrNotAuthorized
+	}
+
+	userClient, err := getClientFromToken(cfg, token)
```

Comment:
Is the previous HTTPS check for certificate auth and this is handling token auth -- but both are HTTPS connections?

Reply:
Does anything need the HTTP connection?  Can we limit it to HTTPS only?

Reply:
I think it's good to have an HTTP option in case you want a proxy to handle the TLS. In ACM, we never will though.

Reply:
I agree to both `http` and `https`. We have CSV endpoint too. I expect individual users would use CSV endpoint using Curl